07_boolean_2/main.go

```diff
@@ -15,56 +15,283 @@ E.g.:
+// "invalid grade".
+// - If the average of the grades is greater than or equal to 90
+//   - If no grade is below 80, return "A+"
+//   - If no grade is below 70, return "A"
+//   - If any grade is below 70, return "A-"
+//
+// - If the average of the grades is greater than or equal to 80, return
+// "B"
+//   - If no grade is below 80, return "B+"
+//   - If no grade is below 70, return "B"
+//   - If any grade is below 60, return "B-"
+//
+// - If the average of the grades is greater than or equal to 70, return
+// "C"
+//
+//   - If no grade is below 70, return "C+"
+//
+//   - If no grade is below 60, return "C"
+//
+//   - If any grade is below 60, return "C-"
+
+// if grade.. > 100 || grade.. < 0 return invalid
+func gradeValid(grade1 int, grade2 int, grade3 int, grade4 int, grade5 int) bool {
+	if grade1 < 0 || grade1 > 100 {
+		return false
+	}
+
+	if grade2 < 0 || grade2 > 100 {
+		return false
+	}
+
+	if grade3 < 0 || grade3 > 100 {
+		return false
+	}
+
+	if grade4 < 0 || grade4 > 100 {
+		return false
+	}
+
+	if grade5 < 0 || grade5 > 100 {
+		return false
+	}
+
+	return true
+}
+
+// check if all grades are >= x
+func allGradesOver(grade1 int, grade2 int, grade3 int, grade4 int, grade5 int, limit int) bool {
+	if grade1 < limit {
+		return false
+	}
+
+	if grade2 < limit {
+		return false
+	}
+
+	if grade3 < limit {
+		return false
+	}
+
+	if grade4 < limit {
+		return false
+	}
+
+	if grade5 < limit {
+		return false
+	}
+
+	return true
+}
+
+// check if grade avg >= limit
+func gradeAvgOver(grade1 int, grade2 int, grade3 int, grade4 int, grade5 int, limit int) bool {
+	var avg int = (grade1 + grade2 + grade3 + grade4 + grade5) / 5
+	return avg >= limit
+}
+
+func calcGrade(grade1 int, grade2 int, grade3 int, grade4 int, grade5 int) string {
+	if !gradeValid(grade1, grade2, grade3, grade4, grade5) {
+		return "invalid grade"
+	}
+
+	if gradeAvgOver(grade1, grade2, grade3, grade4, grade5, 90) {
+		if allGradesOver(grade1, grade2, grade3, grade4, grade5, 80) {
+			return "A+"
+		}
+		if allGradesOver(grade1, grade2, grade3, grade4, grade5, 70) {
+			return "A"
+		}
+		if !allGradesOver(grade1, grade2, grade3, grade4, grade5, 70) {
+			return "A-"
+		}
+	}
+
+	if gradeAvgOver(grade1, grade2, grade3, grade4, grade5, 80) {
+		if allGradesOver(grade1, grade2, grade3, grade4, grade5, 80) {
+			return "B+"
+		}
+		if allGradesOver(grade1, grade2, grade3, grade4, grade5, 70) {
+			return "B"
+		}
+		if !allGradesOver(grade1, grade2, grade3, grade4, grade5, 60) {
+			return "B-"
+		}
+	}
+
+	if gradeAvgOver(grade1, grade2, grade3, grade4, grade5, 70) {
+		if allGradesOver(grade1, grade2, grade3, grade4, grade5, 70) {
+			return "C+"
+		}
+		if allGradesOver(grade1, grade2, grade3, grade4, grade5, 60) {
+			return "C"
+		}
+		if !allGradesOver(grade1, grade2, grade3, grade4, grade5, 60) {
+			return "C-"
+		}
+	}
+	return "grade calc logic not yet implemented for these grades"
+}
```

Comment:
Added this, as I ran into grade combos not being inside my conditions (not being invalid either).